More RISC-V CHERI PTE bits

nwf opened the pull request:
As per the ongoing Sail work at https://github.com/CTSRD-CHERI/sail-cheri-riscv/pull/18.

arichardson commented on target/riscv/cpu_helper.c:
```diff
@@ -622,15 +631,24 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
 
+#if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
+            if (access_type == MMU_DATA_CAP_STORE) {
+                /* SCM and not SC transitions to SC and not SCM */
```
I initially struggled to parse that sentence. Maybe add parens?

arichardson commented on target/riscv/cpu_bits.h:
```diff
@@ -348,6 +348,11 @@
+
+#define CCSR_ENABLE         0x1
+#define CCSR_DIRTY          0x2
+#define CCSR_GCLGS          0x4
```
I feel like these could with a comment expanding the abbreviation?

arichardson commented on target/riscv/cpu_helper.c:
```diff
@@ -688,8 +706,24 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
+                if (pte & PTE_LCM) {
+                    // Cap-loads checked against GCLG in CCSR using PTE_U
+
+                    target_ulong gclgmask = (pte & PTE_U)
```
Indentation looks a bit odd to me, but maybe that's QEMU's style.

nwf replied:
I'm not sure what qemu's style is for this case, where the ? and : but not the third segment fit on one line.  I moved the whole expression down to the next line.

arichardson reviewed (commented):
Looks good to me

bsdjhb commented on target/riscv/cpu_helper.c:
```diff
@@ -622,15 +631,24 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
-        } else if (access_type == MMU_DATA_CAP_STORE && !(pte & PTE_SC)) {
-            /* SC inhibited */
+        } else if (access_type == MMU_DATA_CAP_STORE &&
+                   !(pte & PTE_SC) && !(pte & PTE_SCM)) {
```
Another way you could write this is:

```
(pte & (PTE_SC | PTE_SCM) == 0)
```

I find that this style can be a bit clearer when doing things like:

```
(pte & (PTE_SC | PTE_SCM) == PTE_SC)
```

but I don't have a strong opinion.

bsdjhb commented on target/riscv/cpu_helper.c:
```diff
@@ -622,15 +631,24 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
 
+#if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
+            if (access_type == MMU_DATA_CAP_STORE) {
+                /* (SCM and not SC) transitions to (SC and not SCM) */
```
So I don't see a check for both bits being set.  If that is reserved, we should be checking that above and returning some other fault instead?  (What fault is defined for invalid settings of PTE bits?)

Similarly for invalid combinations of the load cap bits.

jrtc27 replied:
Invalid setting of PTE bits is normally the page fault corresponding to the instruction type. I don't know if that means we'd want a page fault or a capability page fault if the CHERI-specific bits were the issue.

bsdjhb replied:
Hmmm.  x86 has a special code for reserved bits being set (but it also uses a single PF# vector).  I suspect we want to raise the relevant CHERI fault which would happen here by just returning `TRANSLATION_FAIL_CHERI`.

nwf replied:
I'm torn between raising a CHERI-specific fault or following along with the upstream Sail that maps reserved bit combinations to the PTW_Invalid_PTE internal result (and then on to the generic page fault vector).  See

https://github.com/rems-project/sail-riscv/blob/3026c56e7c2ec5e33a4e9f7e54323b9f8ad6f3aa/model/riscv_pte.sail#L37

https://github.com/rems-project/sail-riscv/blob/3026c56e7c2ec5e33a4e9f7e54323b9f8ad6f3aa/model/riscv_vmem_sv48.sail#L29-31

https://github.com/rems-project/sail-riscv/blob/3026c56e7c2ec5e33a4e9f7e54323b9f8ad6f3aa/model/riscv_ptw.sail#L35

nwf replied:
I added checks that raise the existing `TRANSLATE_FAIL` code; if anyone feels strongly that this should be `TRANSLATE_CHERI_FAIL`, I'll flip to that.

jrtc27 commented on target/mips/cp0_helper.c:
```diff
@@ -1241,9 +1241,7 @@ void helper_mtc0_entryhi(CPUMIPSState *env, target_ulong arg1)
 #if defined(TARGET_CHERI)
-    mask |= (1UL << CP0EnHi_CLGU)
-            | (1UL << CP0EnHi_CLGS)
-            | (1UL << CP0EnHi_CLGK);
+    mask |= CP0EnHi_CLG_MASK;
 #endif
```
No need for the `#if` any more

jrtc27 commented on target/riscv/cpu_bits.h:
```diff
@@ -363,6 +363,11 @@
 #define CSR_SCCSR           0x9C0
 #define CSR_MCCSR           0xBC0
+
+#define CCSR_ENABLE         0x1
+#define CCSR_DIRTY          0x2
```
These need some kind of prefix I think, perhaps X given they exist for all the modes.

jrtc27 commented on target/riscv/cpu_bits.h:
```diff
@@ -363,6 +363,11 @@
+
+#define CCSR_ENABLE         0x1
+#define CCSR_DIRTY          0x2
+#define CCSR_GCLGS          0x4 /* Global Capability Load Generation, Super */
+#define CCSR_GCLGU          0x8 /* Global Capability Load Generation, User */
```
```suggestion
#define SCCSR_GCLGS          0x4 /* Global Capability Load Generation, Super */
#define SCCSR_GCLGU          0x8 /* Global Capability Load Generation, User */
```

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -681,8 +696,24 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
+                }
+            } else {
+                if (pte & PTE_CRM) {
+                    // Cap-loads checked against GCLG in CCSR using PTE_U
```
```suggestion
                    /* Cap-loads checked against GCLG in CCSR using PTE_U */
```
C++-style comments plague the CHERI parts but upstream style is to use C-style ones.

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -681,8 +696,24 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
+                    target_ulong gclgmask =
+                            (pte & PTE_U) ? CCSR_GCLGU : CCSR_GCLGS;
+
+                    if (!(env->sccsr & gclgmask) != !(pte & PTE_CRG)) {
```
I know what you're trying to do but it's hard to understand mentally. I'd suggest shifting and masking the low bit instead.

jrtc27 commented on target/riscv/csr.c:
```diff
@@ -1223,24 +1219,44 @@ static int read_ccsr(CPURISCVState *env, int csrno, target_ulong *val)
+    }
+
+    /* Take the GCLG bits from the store and update state bits */
+    env->sccsr = set_field(env->sccsr, gclgmask, get_field(val, gclgmask));
```
MCCSR?

jrtc27 commented on target/riscv/monitor.c:
```diff
@@ -64,6 +64,42 @@ static void print_pte(Monitor *mon, int va_bits, target_ulong vaddr,
+#if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
+    char cr, cw;
+
+    switch (attr & (PTE_CW | PTE_CD)) {
+    case 0: // trap
+        cw = '-';
+        break;
+    case PTE_CD:
+    case PTE_CW | PTE_CD: // store OK
+        cw = 'S';
+        break;
+    case PTE_CW: // store will dirty
+        cw = 's';
+        break;
+    }
```
This should just be handled like W and D now with separate output for each.

nwf replied:
I did the same for the CR{,M,G} bits, too.

jrtc27 commented on target/riscv/monitor.c:
```diff
@@ -112,7 +147,7 @@ static void walk_pte(Monitor *mon, hwaddr base, target_ulong start,
         paddr = (hwaddr)(pte >> PTE_PPN_SHIFT) << PGSHIFT;
 #ifdef TARGET_CHERI
-        attr = pte & (PTE_LC | PTE_SC | 0xff);
+        attr = pte & (PTE_CR | PTE_CW | PTE_CD | PTE_CRM | PTE_CRG | 0xff);
```
Did this work on RV32 before? It looks like it shouldn't (or maybe just silently truncates so it's fine?)...

nwf-msr committed:
CHERI riscv: 2-layer PTW cap-load caveats only from guest tables

Don't yet let the hypervisor play the load-side trap game, but it's
fairly apparent how we might go about it.

nwf-msr committed:
CHERI riscv: CapDirty support

nwf-msr committed:
CHERI riscv: guard against reserved PTE bit configurations

nwf-msr committed:
CHERI riscv: PTE load-side caveats for non-LCG behaviors

nwf-msr committed:
CHERI riscv: rename PTE SC/LC to CW/CR

More obviously parallels the existing W/R bits.

nwf-msr committed:
CHERI MIPS: don't clobber CLG bits in MMU fault

When raising a MMU exception, we were zeroing out the CLG bits from CP0
EntryHi, which was, understandably, making software sad.

Fixes 7c74ddc17a9259eb1a47e54175caa0d11eaf0da2
and f0feea1d13cbf14dff38a1757d7e0d412ec66eda

nwf-msr committed:
CHERI riscv: define additional new PTE bits

nwf-msr committed:
CHERI riscv: expose new PTE bits in monitor

While here, guard the use of all the upper metadata bits with
`!defined(TARGET_RISCV32)`

nwf-msr committed:
CHERI riscv: cap-load generation support

This makes [ms]ccsr writable, but only the GCLG bits are updated.

Co-Authored-By: Jessica Clarke <jrtc27@jrtc27.com>

nwf commented:
Rebasing.  Assuming Jenkins doesn't find any problems and nobody screams, I'll merge.

arichardson commented on target/riscv/csr.c:
```diff
@@ -1237,19 +1233,34 @@ static int read_ccsr(CPURISCVState *env, int csrno, target_ulong *val)
+    }
+
+    /* Take the GCLG bits from the store and update state bits */
+    env->mccsr = set_field(env->mccsr, gclgmask, get_field(val, gclgmask));
```
Do we need another bit somewhere determining whether S-mode is allowed to write?

jrtc27 replied:
No, they're like [ms]status.SIE. But they should really have S prefixes on all of them to make that clear.

arichardson reviewed (approved):
Sorry about the conflict.

arichardson commented:
Test failure is my fault. Will fix shortly.

nwf commented:
Renamed the fields as @jrtc27 suggested.

arichardson reviewed (approved):
I haven't been following the discussion regarding these bits closely. QEMU code looks good.
Will defer functional review to @jrtc27 / @bsdjhb

nwf-msr committed:
CHERI riscv: optionally trap on PTE_{A,D,CD} clear

These are `#if 0`-ed out, but has proven useful in testing, as our FPGA cores
trap rather than atomically update the PTE.

nwf commented:
Could I get some eyes (esp. @jrtc27's and/or @bsdjhb's) so we could merge this PR?  Ben Cole would also like to be using it and carrying branches locally is only so much fun.

nwf commented:
@LawrenceEsswood has been spending a lot of quality time with QEMU of late and so may also have opinions here.

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -362,6 +362,10 @@ void riscv_cpu_set_mode(CPURISCVState *env, target_ulong newpriv)
 #endif
 
+#ifndef RISCV_PTE_TRAPPY
+#define RISCV_PTE_TRAPPY 0 /* any subset of PTE_A | PTE_D | PTE_CD */
```
D and A have to go together per the spec, and I don't think it ever makes sense to have CD trap but not D/A.

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -583,6 +587,15 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
+            /* Reserved CHERI-extended PTE flags: no CR but CRG */
+            return TRANSLATE_CHERI_FAIL;
+        } else if ((pte & (PTE_CR | PTE_CRM | PTE_CRG))
+                        == (PTE_CR | PTE_CRG)) {
```
This formatting seems odd; too indented, and probably should have the == on the line above?

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -603,19 +621,36 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
+                    access_type == MMU_DATA_CAP_STORE) && !(pte & PTE_D)) {
+            /* PTE not marked as dirty */
+            return TRANSLATE_FAIL;
+#endif
```
If we want this, I'd put it below the X check, then it's closer to the sail in precedence

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -594,6 +607,11 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
+        } else if (!(pte & PTE_A)) {
+            /* PTE not marked as accessed */
+            return TRANSLATE_FAIL;
+#endif
```
I'd put this below permissions to match sail

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -603,19 +621,36 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
             target_ulong updated_pte = pte | PTE_A |
                 (access_type == MMU_DATA_STORE ? PTE_D : 0);
 
+#if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
+            if (access_type == MMU_DATA_CAP_STORE) {
+                updated_pte |= PTE_D | PTE_CD;
+            }
+#endif
```
Probably better to include this in the ternary above; bit ugly but ensures there's only one way of doing things

nwf replied:
I moved to a `switch`/`case` instead; does that work for you?

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -673,8 +708,25 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
+                if (pte & PTE_CRM) {
+                    /* Cap-loads checked against [SU]GCLG in CCSR using PTE_U */
+                    target_ulong gclgmask =
+                            (pte & PTE_U) ? SCCSR_UGCLG : SCCSR_SGCLG;
```
Too indented

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -905,6 +957,31 @@ static int riscv_cpu_tlb_fill_impl(CPURISCVState *env, vaddr address, int size,
+             * able to force tag clearing or trapping.  In decreasing priority
+             * order, if the tag comes back set...
+             *
+             *  - If the hypervisor says trap, raise a hypervisor (HS) fault
+             *
+             *  - If the guest says trap, raise a supervisor (VS/S) fault.
```
Other way round, no? Guest translation happens first, and host translation only happens if that succeeds, so the same should be true for tag-dependent traps due to PTEs? Remembering which is going to be annoying though... but Lawrence has probably had to do that for Morello?

jrtc27 commented on target/riscv/csr.c:
```diff
@@ -1347,19 +1343,34 @@ static int read_ccsr(CPURISCVState *env, int csrno, target_ulong *val)
+    ccsr = set_field(ccsr, XCCSR_ENABLE, cpu->cfg.ext_cheri);
+    ccsr = set_field(ccsr, XCCSR_DIRTY, 1); /* Always report dirty */
+
+    /* Expose state bits (GCLG) only above U mode */
+    if (env->priv != PRV_U)
+        ccsr |= env->mccsr;
```
This doesn't make sense. This would expose the bits in uccsr if you read it from S-mode or M-mode. Also, currently Toooba only has then in sccsr; should they be mirrored in mccsr (like this does) or only in sccsr? I feel Toooba's approach makes more sense? Especially once you start to consider a hccsr for H, then it looks more like satp+hgatp.

jrtc27 commented on target/riscv/csr.c:
```diff
@@ -1347,19 +1343,34 @@ static int read_ccsr(CPURISCVState *env, int csrno, target_ulong *val)
+    static const target_ulong gclgmask = (SCCSR_SGCLG | SCCSR_UGCLG);
+
+    /* U mode writes refused */
+    if (env->priv == PRV_U) {
```
Again, by CSR not by current mode, this lets M/S-mode write to uccsr

jrtc27 commented on target/riscv/csr.c:
```diff
@@ -1683,8 +1694,9 @@ static riscv_csr_operations csr_ops[CSR_TABLE_SIZE] = {
-    // CHERI CSRs: For now we alway report the same values and don't allow
-    // turning off any of the bits
+    // CHERI CSRs: For now we always report enabled and dirty and don't support
+    // turning off CHERI.  The global capability load generation bits in here
+    // can be written by S or higher.
```
Comment will need updating to reflect above change requests

jrtc27 commented on target/riscv/monitor.c:
```diff
@@ -80,8 +80,12 @@ static void print_pte(Monitor *mon, int va_bits, target_ulong vaddr,
-                   , attr & PTE_LC ? 'L' : '-'
-                   , attr & PTE_SC ? 'S' : '-'
+                   ,
+                   attr & PTE_CW  ? 'W' : '-',
+                   attr & PTE_CR  ? 'R' : '-',
+                   attr & PTE_CD  ? 'D' : '-',
+                   attr & PTE_CRM ? 'M' : '-',
+                   attr & PTE_CRG ? 'G' : '-'
```
Other way round makes more sense / is consistent with the non-CHERI ones

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -606,15 +625,40 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
             return TRANSLATE_CHERI_FAIL;
+#endif
+#endif
+#if RISCV_PTE_TRAPPY & PTE_A
+        } else if (!(pte & PTE_A)) {
+            /* PTE not marked as accessed */
+            return TRANSLATE_FAIL;
 #endif
```
A comes before D/CD in my head

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -606,15 +625,40 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
-            target_ulong updated_pte = pte | PTE_A |
-                (access_type == MMU_DATA_STORE ? PTE_D : 0);
+            target_ulong updated_pte = pte | PTE_A;
+            switch(access_type) {
```
Space

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -606,15 +625,40 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
+            switch(access_type) {
+#if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
+            case MMU_DATA_CAP_STORE:
+                updated_pte |= PTE_CD | PTE_D; /* as if DATA_STORE, too */
+                break;
```
Why not fall through?

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -905,6 +966,32 @@ static int riscv_cpu_tlb_fill_impl(CPURISCVState *env, vaddr address, int size,
+             * able to force tag clearing or trapping.  In decreasing priority
+             * order, if the tag comes back set...
+             *
+             *  - If the guest says trap, raise a supervisor (VS/S) fault.
+             *
+             *  - If the guest translation succeeds but the hypervisor says
+             *    trap, raise a hypervisor (HS) fault
+             *
+             *  - If either layer says clear, clear the tag.
```
Thinking through more, I wonder what should happen if one says clear and the other says trap.. intuitively I'd say if the host says clear but the guest says trap that shouldn't trap, but if the host says trap and the guest says clear it should trap? Hm, which maybe suggests your view of the world was right, as the loaded data flows in the reverse direction to the order you translate (and the order you store). I wonder what Morello does... regardless, it's not specified, so this doesn't need to be completely thought through yet, just an outline of something seemingly sensible.

nwf replied:
I dropped the "priority order" verbiage and just gave a table of something that seems sensible.  I think "guest says clear host says trap" probably doesn't result in a trap: the guest specifically avoided loading a capability, after all.

jrtc27 replied:
I think it depends what order you have guest and host traps in; if host traps take precedence, then I think conceptually there has been a tagged capability loaded through the host's PTE and so you should trap, even if the tag is then squashed further up. But leaving precedence for later sounds good to me.

jrtc27 commented on target/riscv/csr.c:
```diff
@@ -1347,19 +1343,33 @@ static int read_ccsr(CPURISCVState *env, int csrno, target_ulong *val)
-    return -1;
+    static const target_ulong gclgmask = (SCCSR_SGCLG | SCCSR_UGCLG);
+
+    switch(csrno) {
```
Space

jrtc27 replied:
Still missing

arichardson commented on target/riscv/csr.c:
```diff
@@ -1347,19 +1343,38 @@ static int read_ccsr(CPURISCVState *env, int csrno, target_ulong *val)
+    case CSR_SCCSR: {
+        static const target_ulong gclgmask = (SCCSR_SGCLG | SCCSR_UGCLG);
+        /* Take the GCLG bits from the store and update state bits */
+        env->sccsr = set_field(env->sccsr, gclgmask, get_field(val, gclgmask));
```
Hmm I wonder if we need to flush the QEMU TLB after a write that changes the value here? Otherwise we might end up with stale/missing PAGE_LC_TRAP on the pages?

jrtc27 replied:
Is the same thing not missing from the existing MIPS EntryHi code? It's flushing on ASID changes but not CLGx changes.

nwf-msr committed:
CHERI riscv: flush TLB on ccsr write

nwf commented:
@jrtc27: Are there still changes you want me to make?

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -607,14 +626,45 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
+#if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
+            case MMU_DATA_CAP_STORE:
+                updated_pte |= PTE_CD;
+		/* FALLTHROUGH */
```
Indentation

jrtc27 commented on target/riscv/csr.c:
```diff
@@ -1333,12 +1333,8 @@ static int write_pmpaddr(CPURISCVState *env, int csrno, target_ulong val)
 
 #ifdef TARGET_CHERI
-
 // See Capability Control and Status Registers (CCSRs) in CHERI ISA spec
```
Hm, either this comment should be grouped with the one below (which should also not have a blank line after it) or it should move to where XCCSR_* are defined

nwf replied:
I've moved it to the `XCCSR_*` definitions and pruned the second comment.

nwf-msr committed:
CHERI MIPS: also flush qemu TLB on GCLG change

Thanks to @jrtc27 for pointing out its omission.

arichardson reviewed (approved):
@jrtc27 can we merge this?
